### PR TITLE
Webcrypto, v2 paste format, zlib compression

### LINKED ISSUE #28: Switch to  Web Crypto API
Switch to  [Web Crypto API](https://www.w3.org/TR/WebCryptoAPI/) and provide sjcl only as a fallback.

[Browser support is good](http://caniuse.com/cryptography).

### LINKED ISSUE #49: DB IDs: Get rid of MD5
I am surprised to see how much [MD5 is used in PrivateBin](https://github.com/PrivateBin/PrivateBin/search?utf8=%E2%9C%93&q=md5&type=Code).

Especially why on hell is the the paste id derived from the (presumably) plain text content?
I mean it may be difficult but it basically means that 

And also [this comment](https://github.com/PrivateBin/PrivateBin/blob/002046cc6283ce62b2850a70f29f096cf29a66c8/lib/model/abstract.php#L106-L107) is really ridiculous:

> // We just want a small hash to avoid collisions:
> // Half-MD5 (64 bits) will do the trick

Wait, no! MD5 is not collision resistant at all and fundamentally broken.

I mean completely irrespective of the security impact of MD5 used there, there should not be a single MD5 hashing in a security product. Even if there would be no impact (which I highly doubt) it is a bad image...

### LINKED ISSUE #74: SJCL use - scrutinization
Currently we [hash](https://github.com/PrivateBin/PrivateBin/blob/master/js/privatebin.js#L500) the user-provided password with SHA-256 and converts it to hex. This try to improve the strength of the password may be bad as it creates a new - untested - algorithm and may only make it worse. That's the reason why you should not [home-brew](https://security.stackexchange.com/questions/25585/is-my-developers-home-brew-password-security-right-or-wrong-and-why) algorithms by combining them which can lead to unespected results.
I always like [this example here](https://security.stackexchange.com/questions/33531/why-improvising-your-own-hash-function-out-of-existing-hash-functions-is-so-bad#answer-33550) where bcrypt and scrypt (two secure hash functions) are combined:

> Another potential issue with combining bcrypt and scrypt like this is that there has been very little study into how the two interact. As such, we don't know if there are any weird cases that can cause problems. As a more obvious example, take the one time pad. We compute c=m^k for some message m and some equally long perfectly random key k, and we get perfect security. So let's do it twice, for even more security! That gives us c=m^k^k... oh, wait, that just gives us m. So because we didn't take the time to properly understand how the internals of the system worked, we ended up with a real security vulnerability. Obviously it's more complicated in the case of KDFs, but the same principle applies.

So I wonder whether sjcl does not already stretch and/or hash the password internally, so that this hashing is actually insecure or at least useless.

### LINKED ISSUE #260: Empty paste
@tobast posted on https://github.com/sebsauvage/ZeroBin/issues/114:

>Hello,
>
>In some cases, a (non-empty) paste yields an empty paste link. For instance, when trying to paste this text, ie.
>
> ```
> 
> 1 subgoal
> 
> inv : Assert
> expr : Expr
> sBody : Instr
> deduction : (|- [|inv /\ assertOfExpr expr|] sBody [|inv|])%assert
> IHdeduction : (|= [|inv /\ assertOfExpr expr |] sBody [|inv|])%assert
> mem : Mem
> preInMem : inv mem
> m : Mem
> n : nat
> interpRel : interp (nth_iterate sBody n) (MemElem mem) = CpoElem Mem m
> lastIter : interp (nth_iterate sBody n) (MemElem mem) |=e expr_neg expr
> notLastIter : forall p : nat,
>               p < n -> interp (nth_iterate sBody p) (MemElem mem) |=e expr
> isWhile : interp (while expr sBody) (MemElem mem) =
>           interp (nth_iterate sBody n) (MemElem mem)
> 
> ======================== ( 1 / 1 )
> conseq_or_bottom inv (interp (nth_iterate sBody n) (MemElem mem))
> ```
> 
> the paste is first rendered as non-empty, but following the given link will result in an empty paste…

### LINKED ISSUE #328: Pastes get truncated
## Steps to reproduce

1. Create a paste of ~1.2mb

### What happens

When copy pasting the link into a new browser window, the paste is truncated to a length of 604K.

### What should happen

The paste should not be truncated.

## Basic information

**Server OS:** Gentoo, but PrivateBin runs in docker

**Webserver:** [jwilder/nginx-proxy](https://github.com/jwilder/nginx-proxy) as front proxy with the docker container of PrivateBin

**Browser:** tried with firefox 60.0.2 and chrome 67.0.3396.87

**PrivateBin version:** `privatebin/nginx-fpm-alpine:1.1.1`

The size limit in the config is set to a high value (8mb) and the nginx front proxy is configured with `client_max_body_size 0;` to allow arbitrary big body size.

### PR DESCRIPTION
**Pull request by elrido**
This PR fixes #28, fixes #49, fixes #74, fixes #193, fixes #260, fixes #328.

## Changes
* Removes the following libraries:
  * SJCL.js -> replaced by browser WebCrypto API
  * Base64.js -> replaced by browser atob / btoa functions
  * rawdeflate.js -> replaced by zlib.c, crosscompiled to a WASM module (note that rawinflate.js is still present, but only used for v1 paste decryption)
* Frontend able to decrypt v1 and v2 pastes and comments and URLs, newly created pastes and comments are v2 only
* Backend able to send v1 and v2 (pre-v1 were already converted on the fly to v1 API) pastes and comments, but accepts only v2 pastes and comments to store.
* Refactored backends data passing methods to all work with PHP arrays instead of PHP objects in order to avoid duplicate conversions from and to JSON strings when passing pastes/comments between controller, data model and data storage.
* Not changed: The database table structure remains unchanged. The default filesystem storage always stores the JSON as received by the API version at the time of storage.

## ToDo
* [x] Test mixed formats - pastes created on one version with comments from v1 as well as v2
* [x] Import unit test for #260 and check if it is solved by zlib as well
* [x] Have peers review the changes, implement suggested improvements
* [x] Review impact on code coverage, supply additional unit tests if severe reduction of coverage discovered

### CODE REVIEW

**Comment by elrido** [comment]
### Browser compatibility

TL;DR: Except for IE, all major browsers support the new features as of September 2017.

Apart from the WebCrypto API, this branch also makes use of some other recent features that break backwards compatibility. For reference, here is a list of features and in which version these got supported by the major browsers:

| Browser/Feature | async/await | Web Crypto API | WebAssembly | ECMAScript 2015 (ES6) | first version to support all features | Release date |
|-----------------|-------------|----------------|-------------|-----------------------|---------------------------------------|--------------|
| Chrome          | 55          | 37             | 57          | 51                    | 57                                    | 2017-03-09   |
| Edge            | 15          | 12             | 16          | 15                    | 16                                    | 2017-09-26   |
| Firefox         | 52          | 34             | 52          | 54                    | 54                                    | 2017-06-13   |
| IE              | no          | 11             | no          | 11                    | none                                  | none         |
| Opera           | 42          | 24             | 44          | 38                    | 44                                    | 2017-03-21   |
| Safari          | 10.1        | 11             | 11          | 10                    | 11                                    | 2017-09-19   |

**Comment by elrido** [comment]
I have found a regression while testing: The nickname in comments is now part of the encrypted content. Previously an icon/vizhash was only generated when a nickname was set. Since this is now not known to the server, the current logic generates an icon for every comment. The client then displays the comments for anonymous comments with the headline of (same as if there was a nickname):
[icon] _Anonymous_ [localized date string]
instead of just
_Anonymous_ [localized date string]

While we could just hide the icon if it's an anon comment to preserve the old behaviour, I think that would be misleading. I would therefore suggest we document this change of behaviour in the release notes and let the server admins decide if they want to turn it off for everyone, if it bothers them.

Both the vizhash and identicon libraries are designed and used as to not to leak the IP, but it is in their nature that the commentators are less anonymous then without these icons. If that is problematic for a particular instance, the icons should be disabled, like we already [document in the configuration wiki](https://github.com/PrivateBin/PrivateBin/wiki/Configuration#icon).

**Comment by rugk** [comment]
> While we could just hide the icon if it's an anon comment to preserve the old behaviour, I think that would be misleading.

:+1: 
Despite of disabling vizhash (which is AFAIK already possible for admins) I also see no reason for _not_ having an avatar for an anonymous comment. (It actually makes sense to identify these comments again.) IMHO, we don't need a setting for this. Let's not overcomplicate it with not needed settings…

**Comment by elrido** [comment]
I have now done further testing on the display of existing version 1 pastes with the new logic in the browser as well as the support of mixed version content (paste in v1, comments in v2). Found and fixed some issues with that.

Also solved a regression from base58 that would have made the new pastes occasionally undecryptable as it's encoding strips NULL bytes from the start of the key.

Have also implemented patches suggested by StyleCI and Scrutinizer.

As indicated in #193, I have set up a test server for further review of the new implementation and testing in more browsers then just the latest Firefox and Chrome:
* Test instance: https://zerobin-test.dssr.ch/
* Version 1 paste with version 1 & 2 comments: https://zerobin-test.dssr.ch/?303f5916ec977d79#7WjeUFTClr6bSxxIt4yKXxY/lSij995DowbY319Lws4=
* Version 2 paste with the two formerly failing cases (#260 & #328), that now work using zlib: https://zerobin-test.dssr.ch/?3f711142eea28794#CrjvhH5TnN9YA89MkF4ZWnFuLQYXq82AgVmJHSFEff68

File: js/privatebin.js [hunk 1]
```diff
@@ -4245,27 +4245,34 @@ jQuery.PrivateBin = (function($, RawDeflate) {
...
-                    CryptTool.decipher(key, password, paste.attachmentname)
-                ]).then((attachment) => {
-                    AttachmentViewer.setAttachment(attachment[0], attachment[1]);
-                });
-                PasteViewer.setFormat(paste.meta.formatter);
-                PasteViewer.setText(pastePlain);
-            } else {
+            let format = '',
+                text   = '';
+            if (paste.hasOwnProperty('ct')) {
```
Comment (rugk) [inline js/privatebin.js]: Refactor to some function like `getPasteVersion` or maybe better `isVersion2Paste` or so?

**Comment by cryptomilk** [comment]
Just looked at the test instance. I really like the features and new dark theme! Is there already documentation for the v2 paste format?

**Comment by elrido** [comment]
To clarify: the test instance used the dark bootstrap theme and has the warning banner, the upload and url shortener features enabled, but the URL shortner requires authentication to work and is usable only by myself for testing. All of these features did not change and are already available in the many of the previous versions, just not enabled by default.

Yes, we did discuss and document the v2 API extensively in the [main issue](https://github.com/PrivateBin/PrivateBin/issues/28#issuecomment-432415707).  We still use JSON-LD, so you can just look at the *.jsonld files found in the js folder, same as before:
* https://github.com/PrivateBin/PrivateBin/blob/webcrypto/js/types.jsonld
* https://github.com/PrivateBin/PrivateBin/blob/webcrypto/js/paste.jsonld
* https://github.com/PrivateBin/PrivateBin/blob/webcrypto/js/pastemeta.jsonld
* https://github.com/PrivateBin/PrivateBin/blob/webcrypto/js/comment.jsonld
* https://github.com/PrivateBin/PrivateBin/blob/webcrypto/js/commentmeta.jsonld

The [API wiki page](https://github.com/PrivateBin/PrivateBin/wiki/Encryption-format#format-version-2-privatebin--13) is slightly out of date regarding the JSON structure, but already explains correctly the encryption.

I'll gladly have a look at your python client and could create a pull request with changes for this. We could also publish it as a project directly in the PrivateBin organization as a reference. I'd later also have to look at the nodejs client I had created and update it accordingly.

File: js/privatebin.js (hunk 1 above)
Comment (elrido) [reply js/privatebin.js]: Sure

**Comment by cryptomilk** [comment]
Then I will look into adding support for v2 in the next days: https://git.cryptomilk.org/projects/cpaste.git/

File: cfg/conf.sample.php [hunk 2]
```diff
@@ -70,7 +70,7 @@
 ; Check the documentation at https://content-security-policy.com/
 ; Note: If you use a bootstrap theme, you can remove the allow-popups from the sandbox restrictions.
 ; By default this disallows to load images from third-party servers, e.g. when they are embedded in pastes. If you wish to allow that, you can adjust the policy here. See https://github.com/PrivateBin/PrivateBin/wiki/FAQ#why-does-not-it-load-embedded-images for details.
-; cspheader = "default-src 'none'; manifest-src 'self'; connect-src *; script-src 'self'; style-src 'self'; font-src 'self'; img-src 'self' data:; media-src data:; object-src data:; Referrer-Policy: 'no-referrer'; sandbox allow-same-origin allow-scripts allow-forms allow-popups allow-modals"
+; cspheader = "default-src 'none'; manifest-src 'self'; connect-src *; script-src 'self' 'unsafe-eval'; style-src 'self'; font-src 'self'; img-src 'self' data:; media-src data:; object-src data:; Referrer-Policy: 'no-referrer'; sandbox allow-same-origin allow-scripts allow-forms allow-popups allow-modals"
```
Comment (rugk) [inline cfg/conf.sample.php]: Arrgh, why eval? For what thing is that needed?
As the name says, one really should not be using it…

Comment (rugk) [reply cfg/conf.sample.php]: Ah okay, read your commit message. So seems to be an issue with Crome… arrgh… https://github.com/WebAssembly/content-security-policy/issues/7

Comment (elrido) [reply cfg/conf.sample.php]: I hadn't noticed developing, as I had CSP and HTTPS turned off in the Janitor container

File: js/privatebin.js (hunk 1 above)
Comment (elrido) [reply js/privatebin.js]: Given this code smell, I have extracted several of such version specific pieces of logic into generic Paste and Comment classes. The version can now be checked via the `v` attribute that gets set to 1 by default and version 2 pastes will overwrite. See c4b84b2b6bd8

**Comment by cryptomilk** [comment]
Congratulations! I will look into the Python client as soon as I have time. Probably after Pfingsten.

**Comment by r4sas** [comment]
I think firstly better update API page, add version-depended [en/de]coding references.

**Comment by elrido** [comment]
@r4sas, thanks for reminding me. I did go through both the [API](https://github.com/PrivateBin/PrivateBin/wiki/API) and [encryption format](https://github.com/PrivateBin/PrivateBin/wiki/Encryption-format) pages one more time and updated them with the latest changes.

**Comment by elrido** [comment]
@r4sas & @cryptomilk - out of curiosity I have looked at one of your python clients and [got it working for the paste creation & deletion](https://github.com/PrivateBin/PrivateBin-Python-Cli). I will now also change the decryption in it and then raise a merge request back to the original repository. Optionally we can also maintain such a client in the main projects organization, if you like.

**Comment by cryptomilk** [comment]
I'm fine with maintaining an upstream client and helping out. However to make it easy for cpaste users I will still provide a singe python file as a cpaste client :-)

I'm on the way to the Libre Graphics Meeting right now. Next week is SambaXP conference so I don't really have time to look into privatebin clients right now. However added really awesome stuff to PrivateBin!

**Comment by r4sas** [comment]
> I will now also change the decryption in it

According to latest commits, you want drop v1 format support fully?

**Comment by elrido** [comment]
That is a good point. Apart from sending, folks may want to fetch with your client as well. Also they may not know what version a certain instance is. Thanks to our JSON-LD files we should be able to implement API version detection and then send the according version.

How about I conclude the "get" / decryption part of the client for v2 format, then it is easier to compare what has changed overall. If I add some version detection to the PrivateBin class in api.py, we could maybe have two classes that inherit some base functionality and have the same API for the encryption/paste format and then the logic in the actions.py would just use the class that matches the servers API for creating/parsing the paste?

**Comment by r4sas** [comment]
About creation - I think removing v1 is not problem. But about "get", in json we have `"v": 2` field, which can be used to detect which class use for decryption.

**Pull request merged**